File: 2020/day22/python/day22.py
#!/usr/bin/env python3
from os.path import dirname
from os.path import realpath
from os.path import join
import time
import pytest
import sys

sys.path.append(join(dirname(realpath(__file__)), *["..", ".."]))


def transform_input(input_):
    # custom transform for the day

    decks, player = [], []
    player_nr = -1
    for inp in input_.splitlines():
        if not inp:
            continue
        if "Player" in inp:
            player_nr += 1
            decks.append(player)
            player = []
        else:
            player.append(int(inp))
    decks.append(player)

    return decks[1:]


def read_input(file_name="../data/input.txt"):
    dir_path = dirname(realpath(__file__))
    with open(join(dir_path, file_name), "r") as f:
        input_ = transform_input(f.read())

    return input_


def combat(p1, p2, recursive=False):

    game_config = set()

    winner = 0
    while True:
        if recursive:
            # Player 1 wins
            if (tuple(p1), tuple(p2)) in game_config:
                return p1, 0
            # Otherwise add game configuration
            game_config.add((tuple(p1), tuple(p2)))

        # Draw card
        c1, c2 = p1[0], p2[0]

        winner = -1
        # Check if we enter a game of recursion
        if c1 <= len(p1[1:]) and c2 <= len(p2[1:]) and recursive:
            _, winner = combat(p1[1 : 1 + c1], p2[1 : 1 + c2], recursive)
        # Game rules: here, player 1 wins
        if (c1 > c2 and winner == -1) or winner == 0:
            p1, p2 = p1[1:] + [p1[0]] + [p2[0]], p2[1:]
        # Player 2 wins
        else:
            p1, p2 = p1[1:], p2[1:] + [p2[0]] + [p1[0]]

        if not p1:
            return p2, 1
        elif not p2:
            return p1, 0


def solve_part1(input_):
    # Get card decks for player 1 and player 2
    p1, p2 = input_
    # Play the combat game
    out, winner = combat(p1, p2)

    return sum(i * v for i, v in enumerate(reversed(out), 1))


def solve_part2(input_):

    # Get card decks for player 1 and player 2
    p1, p2 = input_
    # Play the combat game
    out, winner = combat(p1, p2, recursive=True)

    return sum(i * v for i, v in enumerate(reversed(out), 1))


def main():
    input_ = read_input()

    t0 = time.time()
    part1 = solve_part1(input_)
    time_part1 = round((time.time() - t0) * 1e3)
    print(f"Solution to part one: {part1} (time taken {time_part1}[ms])")

    t0 = time.time()
    part2 = solve_part2(input_)
    time_part2 = round((time.time() - t0) * 1e3)
    print(f"Solution to part two: {part2} (time taken {time_part2}[ms])")


if __name__ == "__main__":
    main()


@pytest.mark.parametrize("input1, output1", [("../data/test_input0.txt", 306)])
def test1(input1, output1):
    assert solve_part1(read_input(input1)) == output1


@pytest.mark.parametrize("input2, output2", [("../data/test_input0.txt", 291)])
def test2(input2, output2):
    assert solve_part2(read_input(input2)) == output2

File: 2020/day22/cpp/src/day22.cpp
#include "2020/day22/cpp/include/day22.hpp"

#include <set>

#include "utils/cpp/include/utils.hpp"
using day22::input_type;
using day22::output_type;

input_type day22::transform_input(std::vector<std::string> &input) {
  input_type decks;
  std::vector<int> player;
  int player_nr = -1;

  for (auto &inp : input) {
    // Skip empty line
    if (inp.length() == 0) {
      continue;
    }

    if (inp.find("Player") != std::string::npos) {
      player_nr += 1;
      decks.push_back(player);
      player.clear();
    } else {
      player.push_back(std::stoi(inp));
    }
  }
  decks.push_back(player);
  decks.erase(decks.begin());

  return decks;
}

input_type day22::read_input(std::string file_name) {
  auto input = utils::read_file(file_name);
  return transform_input(input);
}

std::pair<std::vector<int>, int> day22::combat(std::vector<int> p1,
                                               std::vector<int> p2,
                                               bool recursive) {
  std::set<std::pair<std::vector<int>, std::vector<int>>> game_config;
  int winner = 0;

  while (true) {
    if (recursive) {
      if (game_config.count(std::make_pair(p1, p2))) {
        return std::make_pair(p1, 0);
      }
      game_config.insert(std::make_pair(p1, p2));
    }

    // Draw card
    auto c1 = p1[0], c2 = p2[0];

    winner = -1;
    if (((c1 <= int(p1.size()) - 1) && (c2 <= int(p2.size()) - 1)) &&
        recursive) {
      auto output = combat({p1.begin() + 1, p1.begin() + 1 + c1},
                           {p2.begin() + 1, p2.begin() + 1 + c2}, recursive);
      winner = output.second;
    }
    if (((c1 > c2) && (winner == -1)) || winner == 0) {
      std::vector<int> tmp1 = {p1.begin() + 1, p1.end()};
      tmp1.push_back(*p1.begin());
      tmp1.push_back(*p2.begin());
      std::vector<int> tmp2 = {p2.begin() + 1, p2.end()};
      p1 = tmp1;
      p2 = tmp2;
      winner = 0;
    } else {
      std::vector<int> tmp1 = {p1.begin() + 1, p1.end()};
      std::vector<int> tmp2 = {p2.begin() + 1, p2.end()};
      tmp2.push_back(*p2.begin());
      tmp2.push_back(*p1.begin());
      p1 = tmp1;
      p2 = tmp2;
      winner = 1;
    }

    if (p1.size() == 0) {
      return std::make_pair(p2, 1);
    } else if (p2.size() == 0) {
      return std::make_pair(p1, 0);
    }
  }
}

// Implement part one solution
output_type day22::solve_part1(const input_type &input_) {
  std::vector<int> p1 = input_[0];
  std::vector<int> p2 = input_[1];
  auto output = combat(p1, p2);

  int s = 0;
  for (int i = 0; i < int(output.first.size()); i++) {
    s += output.first[i] * (output.first.size() - i);
  }
  return s;
}

// Implement part two solution
output_type day22::solve_part2(const input_type &input_) {
  std::vector<int> p1 = input_[0];
  std::vector<int> p2 = input_[1];
  auto output = combat(p1, p2, true);

  int s = 0;
  for (int i = 0; i < int(output.first.size()); i++) {
    s += output.first[i] * (output.first.size() - i);
  }
  return s;
}
